dask/array/routines.py

```diff
@@ -332,6 +333,52 @@ def vdot(a, b):
     return dot(a.conj().ravel(), b.ravel())
 
 
+def _shape_minus_axis(shape, axis):
+    newshape = list(shape)
+    newshape.pop(axis)
+    return tuple(newshape)
+
+
+def _chunk_sum(a, axis=None, dtype=None, keepdims=None):
+    # Caution: this is not your conventional array-sum: due
+    # to the special nature of the preceding blockwise con-
+    # traction,  each chunk is expected to have exactly the
+    # same shape,  with a size of 1 for the dimension given
+    # by `axis` (the reduction axis).  This makes mere ele-
+    # ment-wise addition of the arrays possible.   Besides,
+    # the output can be merely reshaped to lose the `axis`-
+    # dimension when keepdims = False
+    if type(a) is list:
+        item_type = type(a[0])
+        if item_type is np.ndarray:
+            out = chunk.sum(a, dtype=dtype, axis=0)
+        else:
+            a = [item.__array__() for item in a]
+            out = chunk.sum(a, dtype=dtype, axis=0)
+            out = item_type(out)
```

Comment:
I'm probably missing something, but why does `chunk.sum` doesn't work here? could we let numpy delegate to the array types? `__array__` might lead to numpy array materialisation right, which can be expensive? afair CuPy doesn't support `__array__`.

Reply:
I agree that there's a problem here.  But I'm not sure how to fix it properly.  

It turns out that `chunk.sum`  works for the latest version of numpy so then this conditional branch is unnecessary.  But the CI tests revealed that earlier versions of numpy raise a `ValueError: setting an array element with a sequence.`

I'll explore further.  Thanks.

Reply:
I forgot to mention &mdash; the specific chunk type that raises the above exception is `dask.array.tests.test_dispatch.EncapsulateNDArray` found defined in dask/array/tests/test_dispatch.py

Reply:
I think you need to add `squeeze` to `EncapsulateNDArray`, see https://github.com/dask/dask/blob/a5aecac8313fea30c5503f534c71f325b1775b9c/dask/array/tests/test_dispatch.py#L84-L87

If we fix that what else is missing?

Reply:
I wanted to do that earlier but hesitated because got the impression the creators wanted the `EncapsulateNDArray` interface to remain minimal as mentioned in the docstring for `WrappedArray` in [dask/dask/array/tests/test_dispatch.py](https://github.com/dask/dask/blob/a5aecac8313fea30c5503f534c71f325b1775b9c/dask/array/tests/test_dispatch.py#L93):
```python
class WrappedArray(np.lib.mixins.NDArrayOperatorsMixin):
    """
    Another mock duck array class (like EncapsulateNDArray), but
    designed to be above Dask in the type casting hierarchy (that is,
    WrappedArray wraps Dask Array) and be even more minimal in API.
    Tests that Dask defers properly to upcast types.
    """
``` 

Perhaps you feel that's not critical?

Reply:
It is my **believe**, that if we go forward with using `squeeze` in `matmul` (which arguable is a pretty common op), adding `squeeze` to that test `WrappedArray` is "justified".

Reply:
True.  So should I add it and revert back to using `numpy.sum` instead of `numpy.add` with `functools.reduce`?  Is there anyone who'd be unhappy with that?

Reply:
I'm +1, I would be interested to see the impact there, also to the other comment: https://github.com/dask/dask/pull/8423#discussion_r761478288